sdks/go/pkg/beam/io/xlang/kafkaio/kafka.go

```diff
@@ -0,0 +1,288 @@
+// different SDK (listed below). During pipeline construction, the Go SDK will
+// need to connect to an expansion service containing information on these
+// transforms in their native SDK.
+//
+// To use an expansion service, it must be run as a separate process accessible
+// during pipeline construction. The address of that process must be passed to
+// the transforms in this package.
+//
+// The version of the expansion service should match the version of the Beam SDK
+// being used. For numbered releases of Beam, these expansions services are
+// released to the Maven repository as modules. For development versions of
+// Beam, it is recommended to build and run it from source using Gradle.
+//
+// Current supported SDKs, including expansion service modules and reference
+// documentation:
+// * Java
+//    - Vendored Module: beam-sdks-java-io-expansion-service
+//    - Run via Gradle: ./gradlew :sdks:java:io:expansion-service:runExpansionService
+//    - Reference Class: org.apache.beam.sdk.io.kafka.KafkaIO
+package kafkaio
+
+// TODO(BEAM-12492): Implement an API for specifying Kafka type serializers and
+// deserializers.
+
+import (
+	"reflect"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/graph"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/typex"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/reflectx"
+)
+
+func init() {
+	beam.RegisterType(reflect.TypeOf((*readPayload)(nil)).Elem())
+	beam.RegisterType(reflect.TypeOf((*writePayload)(nil)).Elem())
+}
+
+type policy string
+
+const (
+	ByteArrayDeserializer = "org.apache.kafka.common.serialization.ByteArrayDeserializer"
+	ByteArraySerializer   = "org.apache.kafka.common.serialization.ByteArraySerializer"
+
+	// ProcessingTime is a timestamp policy that assigns processing time to
+	// each record. Specifically, this is the timestamp when the record becomes
+	// "current" in the reader. Further documentation can be found in Java's
+	// KafkaIO documentation.
+	ProcessingTime policy = "ProcessingTime"
+
+	// CreateTime is a timestamp policy based on the CREATE_TIME timestamps of
+	// kafka records. Requires the records to have a type set to
+	// org.apache.kafka.common.record.TimestampTypeCREATE_TIME. Further
+	// documentation can be found in Java's KafkaIO documentation.
+	CreateTime policy = "CreateTime"
+
+	// LogAppendTime is a timestamp policy that assigns Kafka's log append time
+	// (server side ingestion time) to each record. Further documentation can
+	// be found in Java's KafkaIO documentation.
+	LogAppendTime policy = "LogAppendTime"
+
+	readURN  = "beam:external:java:kafka:read:v1"
+	writeURN = "beam:external:java:kafka:write:v1"
+)
+
+// Read is a cross-language PTransform which reads from Kafka and returns a
+// KV pair for each item in the specified Kafka topics. By default, this runs
+// as an unbounded transform and outputs keys and values as raw byte arrays.
```

Comment:
```suggestion
// as an unbounded transform and outputs keys and values as byte slices.
```
Byte arrays have a different meaning in Go (arrays are all fixed length), and the user would be receiving them as slices.